### PR TITLE
MonadIO instances + redeem

### PR DESCRIPTION
**Pull request by emilypi**
Here are the MonadIO instances available to us, as well as a shameless theft of `redeem` from ZIO

### CODE REVIEW

**Comment by emilypi** [comment]
@fommil I know you have a special way you'dlike to do instances, feel free to reconfigure, or give a run down of what you thinkis best

**Comment by fommil** [comment]
Cool!

In 7.2 the MonadIO is able to lift nested monad transformers, did you look into how it is able to do that?

It might be worthwhile overriding a few more things for performance. Especially map.

That said, I prefer the implicit Monad approach. Would it be hard to change that?

**Comment by emilypi** [comment]
@fommil naw i could change it relatively easily - it just means i get to do less work in the instance creation 😄

**Comment by fommil** [comment]
also, shouldn't this be `LiftIO` not `MonadIO` ?

I don't understand the point of `MonadIO`.

**Comment by emilypi** [comment]
@fommil `LiftIO` is what `cats.effect` renamed `MonadIO`. https://hackage.haskell.org/package/transformers-0.2.2.1/docs/Control-Monad-IO-Class.html

I'd rather be at parity with Haskell than further their poor naming conventions

**Comment by fommil** [comment]
we have prior art https://github.com/scalaz/scalaz/blob/series/7.3.x/effect/src/main/scala/scalaz/effect/LiftIO.scala (going back to 2011, at least)

**Comment by fommil** [comment]
I don't think this needs an implicit `Monad` to implement.

I think the `E` should be on the typeclass rather than the method.

**Comment by emilypi** [comment]
@fommil It looks like they made `LiftIO` `MonadIO`, and `MonadIO` `Monad with LiftIO`. I don't know about that, since `MonadIO` requires monad laws to make sense as a typeclass. There really isn't a semistructure here to abstract out. It's one method

**Comment by emilypi** [comment]
@fommil alright I've switched it up so we're no longer using inheritance, but now the implicits aren't being found for each transformer appropriately. Would this have something to do with the new wait we're making instances discoverable? This should compile.

**Comment by fommil** [comment]
if you want laws, you'll need to extend `Monad` and add them using the scalaz style, e.g. see how `Monad` does it from the `Bind` laws.

**Comment by fommil** [comment]
```
[error] MonadIOInstances.scala:16:25: type parameter E defined in method liftIO shadows type E defined in method ioMonadIO. You may want to rename your type parameter, or possibly remove it.
[error]     override def liftIO[E, A](io: IO[E, A])(implicit M: Monad[IO[E, ?]]): IO[E, A] = io
[error]                         ^
```

**Comment by emilypi** [comment]
Will do. I'll get this fixed up by tonight.

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+abstract class MonadIOInstances {
+  import Isomorphism._
+
+  def fromIso[F[_], G[_]](D: F <~> G)(implicit E: MonadIO[G]): MonadIO[F] =
+    new IsomorphismMonadIO[F, G] {
+      override def G: MonadIO[G] = E
+      override def iso: F <~> G  = D
+    }
+
+  implicit def ioMonadIO[E]: MonadIO[IO[E, ?]] = new MonadIO[IO[E, ?]] {
```
Comment (fommil) [inline ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: move this to a parent `abstract class` and add an `implicit val` for `Task` (caches the instance, and has higher implicit priority)

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      override def liftIO[E, A](io: IO[E, A])(implicit M: Monad[OptionT[F, _]]): OptionT[F, A] =
+        OptionT(F.liftIO(io).map(Some.apply))
+    }
+
+  implicit def theseTMonadIO[F[_], E](implicit F: MonadIO[F], F1: Functor[F]): MonadIO[TheseT[F, E, ?]] =
+    new MonadIO[TheseT[F, E, ?]] {
+      override def liftIO[E, A](io: IO[E, A])(implicit M: Monad[TheseT[F, E, ?]]): TheseT[F, E, A] =
+        TheseT(F.liftIO(io).map(\&/.That.apply))
+    }
+
```
Comment (fommil) [inline ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: missing a bunch of stuff that's in https://github.com/scalaz/scalaz/blob/series/7.3.x/effect/src/main/scala/scalaz/effect/LiftIO.scala e.g. kleisli

Comment (emilypi) [reply ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: `ReaderT` _is_ `Kleisli` per its alias. Supporting `Region` doesn't seem right, considering it's use case being deprecated in favor of this project (or zio in the future). Also, do we really want to continue the `ListT`/`StreamT` farce? It's been deprecated from Haskell for years and both are entirely unlawful.

Comment (emilypi) [reply ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: https://hackage.haskell.org/package/transformers-0.5.4.0/docs/Control-Monad-Trans-List.html

Comment (fommil) [reply ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: if it's not deleted from 7.3, it should be. You know you want to.

Comment (emilypi) [reply ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: I dooooooooo 😄

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIO.scala [hunk 3]
```diff
@@ -12,16 +14,33 @@ import scalaz.Monad
  *
  * @tparam M - the monad in which to lift
  */
-trait MonadIO[M[_]] extends Monad[M] {
+trait MonadIO[M[_], E] {
```
Comment (fommil) [inline ioeffect/src/main/scala/scalaz/ioeffect/MonadIO.scala]: missing E in the docs.

You write docs, you create a liability for the next person :stuck_out_tongue: to be wrong.

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  implicit def ioMonadIO[E]: MonadIO[IO[E, ?], E] = new MonadIO[IO[E, ?], E] {
+    override def liftIO[A](io: IO[E, A])(implicit M: Monad[IO[E, ?]]): IO[E, A] = io
+  }
+}
+
+sealed abstract class MonadIOInstances2 {
+  import Isomorphism._
+
+  def fromIso[F[_], G[_], E](D: F <~> G)(implicit E: MonadIO[G, E], M1: Monad[G]): MonadIO[F, E] =
```
Comment (fommil) [inline ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: `fromIso` typically goes on the companion.

If you need access to it from here you can do a self type

`this: MonadIO.type =>`

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    new IsomorphismMonadIO[F, G, E] {
+      override def G: MonadIO[G, E] = E
+      override def M: Monad[G]      = M1
+      override def iso: F <~> G     = D
+    }
+
+  implicit def identityTMonadIO[M[_], E](implicit M: MonadIO[M, E], M1: Monad[M]): MonadIO[IdT[M, ?], E] =
```
Comment (fommil) [inline ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: I don't think these need to go on the lowest priority trait. You can put them into the companion (or highest priority).

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+  implicit def theseTMonadIO[F[_], E0, E](implicit F: MonadIO[F, E], M: Monad[F]): MonadIO[TheseT[F, E0, ?], E] =
+    new MonadIO[TheseT[F, E0, ?], E] {
+      override def liftIO[A](io: IO[E, A])(implicit M1: Monad[TheseT[F, E0, ?]]): TheseT[F, E0, A] =
+        TheseT(F.liftIO(io).map(\&/.That.apply))
+    }
+
+}
+
+abstract class IsomorphismMonadIO[F[_], G[_], E] extends MonadIO[F, E] {
```
Comment (fommil) [inline ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: this also typically goes in the same file as the typeclass in 7.2 and 7.3

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIO.scala (hunk 3 above)
Comment (emilypi) [reply ioeffect/src/main/scala/scalaz/ioeffect/MonadIO.scala]: FFFFFFFFFFF yeah I'll amend the docs. Good catch.

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala (hunk 4 above)
Comment (emilypi) [reply ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: Will just move to companion

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala (hunk 5 above)
Comment (emilypi) [reply ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: 👍

File: ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala (hunk 6 above)
Comment (emilypi) [reply ioeffect/src/main/scala/scalaz/ioeffect/MonadIOInstances.scala]: 👍

**Comment by emilypi** [comment]
@fommil alright, looks like we're good to go on this one

**Pull request merged**